analyzers/rolaguard_base_analyzer/ResourceMeter.py

```diff
@@ -0,0 +1,110 @@
+from db.Models import Device, Gateway
+from datetime import date
+
+
+class ResourceMeter():
+    maw = 0.9 # Moving average weight
+
+    def __init__(self):
+        self.device_stats = {}
+        self.gateway_stats = {}
+        self.last_gc = date.today()
+
+    def __call__(self, asset, packet):
+        if not asset or not packet: return
+        if not packet.f_count or not packet.date: return
+        if packet.m_type in ["JoinRequest", "JoinAccept"]: return
+        
+        packet.uplink = packet.m_type in ["ConfirmedDataUp", "UnconfirmedDataUp"]
+
+        if type(asset) == Device:
+            if asset.id in self.device_stats:
+                self.measure_device_stats(asset, packet)
+                if packet.uplink:
+                    self.device_stats[asset.id]["last_fcount"] = packet.f_count
+                else:
+                    self.device_stats[asset.id]["last_fcount_down"] = packet.f_count
+                self.device_stats[asset.id]["last_date"] = packet.date
+            else:
+                self.device_stats[asset.id] = {
+                    "last_fcount" : packet.f_count if packet.uplink else None,
+                    "last_fcount_down" : packet.f_count if not packet.uplink else None,
+                    "last_date" : packet.date,
+                    "rssi" : {packet.gateway : packet.rssi}
+                }
+        if type(asset) == Gateway:
+            if asset.id in self.gateway_stats:
+                self.measure_gateway_stats(asset, packet)
+            self.gateway_stats[asset.id] = {
+                "last_fcount" : packet.f_count,
+                "last_tmst" : packet.tmst,
+                "last_date" : packet.date
+            }
+
+
+    def measure_device_stats(self, device, packet):
+        if packet.uplink and packet.f_count == self.device_stats[device.id]["last_fcount"]: 
+            return # Repeated uplink packet
+        if not packet.uplink and packet.f_count == self.device_stats[device.id]["last_fcount_down"]: 
+            return # Repeated downlink packet
+
+        device.npackets_up += 1 if packet.uplink else 0
+        device.npackets_down += 1 if not packet.uplink else 0
+
+        if packet.uplink and self.device_stats[device.id]["last_fcount"]:
+            device.last_activity = packet.date
+            device.connected = True
+
+            last_fcount = self.device_stats[device.id]["last_fcount"]
+            count_diff = int(packet.f_count - last_fcount) % (2**16)
+
+            # The counter changed a lot, probably the session was restarted
+            if count_diff > 64 or count_diff < 0:
+                del self.device_stats[device.id]
+                return
+
+            time_diff = (packet.date - self.device_stats[device.id]["last_date"]).seconds / count_diff
```

Comment:
If count_diff is exactly 0, it will occur a runtime error due to a division by zero (this will happen if f_count difference is a multiple of 2**16). May be it's a good idea to consider the count_diff == 0 case in the if condition.

Reply:
Solved in the previous if.